### PR TITLE
feat: walletconnect

### LINKED ISSUE #117: ️🛠 - Initial WalletConnect Setup
### Description

Familiarize yourself with Walletconnect and set it up into our application. Implement an option to connect with a dapp (https://react-app.walletconnect.com/) and add some mock handlers for the events

### Requirements

- [x] Add walletconnect to application
- [x] Add option to connect to dapp
- [x] Mock event handlers
- [x] Create demo project for wallet in walletconnect dashboard

### Checklist

- [X] I have linked the correct milestone and/or project(s)
- [X] I have added the appropriate labels
- [X] I have assigned this task to the correct people

### PR DESCRIPTION
**Pull request by MarkNerdi**
## Summary

1. Enable `walletConnect` feature flag
2. Go to https://react-app.walletconnect.com/
3. Select "Ethereum" and click `Connect`
4. Copy on the top right the connection code
5. Go to wallet and click `Connect` and insert the code
6. Click on `personal_sign` on the webapp, bloom should open a popup

## Changelog

```
- Please write a list of changes
```

### Instructions

> Please describe relevant instructions, configurations, etc. for testing your changes.

...

### Platforms

> Please select any platforms where your changes have been tested.

- **Desktop**
  - [ ] MacOS
  - [ ] Linux
  - [ ] Windows
- **Mobile**
  - [ ] iOS
  - [ ] Android
- **Web**
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Brave
  - [ ] Safari

## Checklist

> Please tick the following boxes that are relevant to your changes.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my changes
- [ ] I have added or modified tests, proving my changes work as intended
- [ ] I have added comments to my code for hard-to-understand areas
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have verified that my latest changes pass CI workflows for testing and linting
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/desktop/components/popups/InitWalletConnectPopup.svelte [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<script lang="ts">
+    import { localize } from '@core/i18n'
+    import { closePopup } from '@desktop/auxiliary/popup'
+    import { pairWithNewApp } from '@lib/walletconnect/utils'
+    import { Button, Text, TextInput, FontWeight, TextType } from 'shared/components'
+
+    let wcUri: string
```
Comment (Tuditi) [inline packages/desktop/components/popups/InitWalletConnectPopup.svelte]:
```suggestion
    let walletConnectUri: string
```

File: packages/desktop/package.json [hunk 2]
```diff
@@ -40,6 +40,7 @@
         "buffer": "^6.0.3",
         "dotenv": "^16.0.3",
         "electron-dl": "^3.5.0",
+        "@walletconnect/web3wallet": "^1.4.0",
```
Comment (Tuditi) [inline packages/desktop/package.json]:
```suggestion
        "@walletconnect/web3wallet": "^1.8.2",
```

File: packages/shared/lib/auxiliary/deep-link/handlers/handleDeepLink.ts [hunk 3]
```diff
@@ -63,7 +64,20 @@ function handleDeepLinkForHostname(url: URL): void {
...
+            handleConnect(url)
+            break
         default:
-            throw new Error(`Unrecognized context '${url.host}'`)
+            throw new Error(`Unrecognized context '${url.hostname}'`)
+    }
+}
+
+function handleConnect(url: URL): void {
+    const wcConnectUri = url.pathname.split('/')[1] + url.search
```
Comment (Tuditi) [inline packages/shared/lib/auxiliary/deep-link/handlers/handleDeepLink.ts]:
```suggestion
    const walletConnectUri = url.pathname.split('/')[1] + url.search
```

File: packages/shared/lib/tests/setup.ts [hunk 4]
```diff
@@ -6,6 +6,9 @@
 import './__mocks__/api.mock'
 import './__mocks__/match-media.mock'
 import './__mocks__/platform.mock'
+import { TextEncoder, TextDecoder } from 'util'
+
+Object.assign(global, { TextDecoder, TextEncoder })
```
Comment (Tuditi) [inline packages/shared/lib/tests/setup.ts]: Maybe a comment why we added a polyfill here would be helpful.

File: packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { JsonRpcResponse } from '@walletconnect/jsonrpc-types'
+
+export function handleEthSendTransaction(
+    id: number,
+    params: unknown,
+    responseCallback: (response: JsonRpcResponse) => void
+): void {
+    if (!params || !Array.isArray(params)) {
+        responseCallback({ id, error: { code: 5000, message: 'Error' }, jsonrpc: '2.0' })
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts]: Seems like this error message could be improved :relaxed:

File: packages/shared/lib/walletconnect/handlers/onSessionRequest.handler.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import { handlePersonalSign } from './personal_sign.handler'
+import { JsonRpcResponse } from '@walletconnect/jsonrpc-types'
+import { Web3WalletTypes } from '@walletconnect/web3wallet'
+import { get } from 'svelte/store'
+import { walletClient } from '../stores'
+
+export function onSessionRequest(event: Web3WalletTypes.SessionRequest): void {
+    const { topic, params, id } = event
+    const { request } = params
+    // const chainId = params.chainId
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/handlers/onSessionRequest.handler.ts]: Add a todo or remove the comment if we don't need it :relaxed:

File: packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    params: unknown,
+    responseCallback: (response: JsonRpcResponse) => void
+): void {
+    if (!params || !Array.isArray(params)) {
+        responseCallback({ id, error: { code: 5000, message: 'Error' }, jsonrpc: '2.0' })
+        return
+    }
+
+    const hexMessage = params[0]
+    // const address = params[1]
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts]: Same in this file.

File: packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts (hunk 5 above)
Comment (Tuditi) [reply packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts]: Additionally, we should rethink the validation. It looks similar to the other handler, so maybe we can de-duplicate this using a Decorator or by moving it to `onSessionRequest`?

File: packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    // const signedMessage = await wallet.signMessage(message)
+
+    const signedMessage = 'hello this is signed'
+    const response = { id, result: signedMessage, jsonrpc: '2.0' }
+
+    openPopup({
+        id: PopupId.Confirmation,
+        props: {
+            title: 'Personal Sign',
+            description: 'Do you wanna sign the following message: ' + message,
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts]: Use a locale here

File: packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    const hexMessage = params[0]
+    // const address = params[1]
+
+    if (typeof hexMessage !== 'string') {
+        responseCallback({ id, error: { code: 5000, message: 'Error' }, jsonrpc: '2.0' })
+        return
+    }
+
+    const message = Converter.hexToUtf8(hexMessage)
+    // sign the message
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts]: Add an issue for this.

File: packages/shared/lib/walletconnect/utils/initializeWalletConnect.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import features from '@features/features'
+import { WALLET_METADATA } from '../constants'
+import { onSessionProposal, onSessionRequest } from '../handlers'
+import { walletClient } from '../stores'
+import { Web3Wallet } from '@walletconnect/web3wallet'
+import { Core } from '@walletconnect/core'
+
+const core = new Core({
+    projectId: process.env.WALLETCONNECT_PROJECT_ID ?? '41511f9b50c46a80cdf8bd1a3532f2f9',
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/utils/initializeWalletConnect.ts]: What does the default value represent? Is it your personal project ID?

File: packages/shared/lib/walletconnect/stores/wallet-client.store.ts [hunk 11]
```diff
@@ -0,0 +1,4 @@
+import { Web3Wallet } from '@walletconnect/web3wallet/dist/types/client'
+import { Writable, writable } from 'svelte/store'
+
+export const walletClient: Writable<Web3Wallet | undefined> = writable(undefined)
```
Comment (Tuditi) [inline packages/shared/lib/walletconnect/stores/wallet-client.store.ts]: Add `getWalletClient()` and use that to get the client :relaxed:

**Review by Tuditi — CHANGES_REQUESTED** [review]
I get the following error when trying to connect:

```
Error: Non conforming namespaces. approve() namespaces chains don't satisfy required namespaces.
      Required: eip155:5
      Approved: 
    at St (index.es.js:4:1)
    at onSessionProposal (onSessionProposal.handler.ts:14:1)
    at o.<anonymous> (initializeWalletConnect.ts:23:1)
    at o.emit (index.es.js:1:1)
    at EventEmitter.onSessionProposal (index.es.js:1:1)
    at EventEmitter.emit (events.js:153:1)
    at ss.onSessionProposeRequest (index.es.js:1:1)
 ```
 This is only shown in the console. Maybe we should also add it to the error log and show a popup?

File: packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts (hunk 5 above)
Comment (MarkNerdi) [reply packages/shared/lib/walletconnect/handlers/eth_sendTransaction.handler.ts]: the expected form (eg. array, obj, string etc) could always be different for each event

File: packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts (hunk 9 above)
Comment (MarkNerdi) [reply packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts]: https://github.com/bloomwalletio/bloom/issues/122

File: packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts (hunk 8 above)
Comment (MarkNerdi) [reply packages/shared/lib/walletconnect/handlers/personal_sign.handler.ts]: this currently is just for mocking. we will have a different kind of popup once we do the real implementation

File: packages/shared/lib/walletconnect/utils/initializeWalletConnect.ts (hunk 10 above)
Comment (MarkNerdi) [reply packages/shared/lib/walletconnect/utils/initializeWalletConnect.ts]: yes, with this we can develop on dapp connection without exposing our real enterprise project

**Review by Tuditi — APPROVED** [review]
LGTM! :fire:

**Pull request merged**